### PR TITLE
[MM-54024] Handle JSON nulls when unmarshalling in api4

### PR DESCRIPTION
**Pull request by Delaney**
#### Summary
Fixes code where `null` JSON values are not handled in the `api4` files.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/24246
Jira https://mattermost.atlassian.net/browse/MM-54024

#### Release Note

```release-note
Handle JSON null value case correctly by also checking pointer is no longer null when unmarshalling to a pointer
```

### CODE REVIEW

**Review by hanzei — CHANGES_REQUESTED** [review]
The changes look fine, but there are some instances that you missed like https://github.com/mattermost/mattermost/blob/6dc472e8ae920512631de74c59a3b79ce97a33f1/server/channels/api4/channel.go#L86-L91

I've searched for them using the following regex: `var (.*) \*`. Maybe that helps you find addition code pieces that need this threadment.

**Review by hanzei — APPROVED** [review]
LGTM :+1:

File: server/channels/api4/team.go [hunk 1]
```diff
@@ -816,7 +816,7 @@ func addTeamMembers(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	var appErr *model.AppError
 	var members []*model.TeamMember
-	if jsonErr := json.NewDecoder(r.Body).Decode(&members); jsonErr != nil {
+	if jsonErr := json.NewDecoder(r.Body).Decode(&members); jsonErr != nil || members == nil {
```
Comment (agnivade) [inline server/channels/api4/team.go:819]: Is this really needed? There's a check which verifies the length of a slice after this.

File: server/channels/api4/cloud.go [hunk 2]
```diff
@@ -399,7 +399,7 @@ func getCloudProducts(c *Context, w http.ResponseWriter, r *http.Request) {
 	if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionSysconsoleReadBilling) {
 		sanitizedProducts := []model.UserFacingProduct{}
 		err = json.Unmarshal(byteProductsData, &sanitizedProducts)
-		if err != nil {
+		if err != nil || sanitizedProducts == nil {
```
Comment (agnivade) [inline server/channels/api4/cloud.go:402]: Same. This is unncessary as well. It will just write `null` as the output which is valid if the input is null.

**Review by agnivade — CHANGES_REQUESTED** [review]
Looks good, but can we add a few unit tests?

**Comment by Delaney** [comment]
@agnivade I'm still writing tests, will be pushing those soon. Also, I'm unable to test on my local setup. Keeping getting this error:

```
panic: cannot read file: open mattermost/server/channels/testlib/testdata/postgres_migration_warmup.sql: no such file or directory
```

Could I get some help with that?

**Comment by agnivade** [comment]
Sure. Can you tell me what have you tried to solve the issue? Maybe you can search the codebase for the error, or set a debugger while running the test and see where it fails?

**Comment by Delaney** [comment]
@agnivade I think, for some reason, `postgres_migration_warmup.sql` can't be found. I'm using a hacky method where I'm executing the SQL directly, and skipping the `PreloadMigrations` method in `testlib/helper.go`

**Comment by Delaney** [comment]
> Looks good, but can we add a few unit tests?

More tests added, ready for review

**Review by agnivade — APPROVED** [review]
Thanks!

**Comment by agnivade** [comment]
@saturninoabril - The e2e [test](https://automation-dashboard.vercel.app/cycle/375233) shows one unrelated failure. What should be the procedure here?

**Comment by saturninoabril** [comment]
True, it's unrelated but it's fixed on master - https://github.com/mattermost/mattermost/pull/24614. This could be updated and may run the test again.

**Pull request merged**